### ISSUE TITLE
Report bugs to porkbun dns api

### ISSUE THREAD

**Issue by qkdreyer** [acmesh-official/acme.sh#3450]
This is the place to report bugs in the [porkbun](https://porkbun.com) DNS API.

If you experience a bug, please report it in this issue.

Thanks!

**Comment by sevengiants** [comment]
thanks, tested and worked fine.

**Comment by GeorgeRT** [comment]
Not sure what to submit but issuing a cert for a subdomain creates TXT record for the domain and hence verify fails.

**Comment by isbjorn** [comment]
@GeorgeRT With this patch I made cert for my 4th-level subdomain:

```
*** /home/foo/dns_porkbun.sh        2021-04-04 21:19:04.628732412 +0000
--- /home/foo/dns_porkbun.sh.working        2021-04-04 23:53:21.227966009 +0000
***************
*** 110,117 ****

      if _porkbun_rest POST "dns/retrieve/$h"; then
        if _contains "$response" "\"status\":\"SUCCESS\""; then
-         _sub_domain="$(echo "$fulldomain" | sed "s/\\.$_domain\$//")"
          _domain=$h
          return 0
        else
          _debug "Go to next level of $_domain"
--- 110,117 ----

      if _porkbun_rest POST "dns/retrieve/$h"; then
        if _contains "$response" "\"status\":\"SUCCESS\""; then
          _domain=$h
+         _sub_domain="$(echo "$fulldomain" | sed "s/\\.$_domain\$//")"
          return 0
        else
          _debug "Go to next level of $_domain"
```

tldr: in function `_get_root` value of `_sub_domain` should be calculated after value of `_domain`

Keep in mind that dns changes may take a long time to propagate. Like 10 minutes or more.

**Comment by GeorgeRT** [comment]
That woks. Thanks.

**Comment by Haarolean** [comment]
The records being added via API have the wrong domain name.
I've issued a cert via
`./acme.sh --issue --domain cronfire.ru --domain *.cronfire.ru --home "." --dns dns_porkbun --challenge-alias le.cronfire.ru`
But the records added are these:
![image](https://user-images.githubusercontent.com/1494347/119234991-0d8ff400-bb39-11eb-8cbd-29294e588217.png)
Names should be 'le.cronfire.ru', not 'le.le.cronfire.ru'
`[Sat May 22 20:03:30 MSK 2021] Checking cronfire.ru for _acme-challenge.le.cronfire.ru`

**Comment by Haarolean** [comment]
Second issue:
After a successful issue there's this error:

```
Sat May 22 20:11:39 MSK 2021] Already success, continue next one.
[Sat May 22 20:11:39 MSK 2021] All success, let's return
[Sat May 22 20:11:39 MSK 2021] Verifying: cronfire.ru
[Sat May 22 20:11:43 MSK 2021] cronfire.ru:Verify error:DNS problem: NXDOMAIN looking up TXT for _acme-challenge.cronfire.ru - check that a DNS record exists for this domain
[Sat May 22 20:11:43 MSK 2021] Removing DNS records.
[Sat May 22 20:11:43 MSK 2021] Removing txt: khSTps6PkRkW3xA-8igZupkNOrqM7e-LiY2rHp3U214 for domain: _acme-challenge.le.cronfire.ru
[Sat May 22 20:11:59 MSK 2021] Removed: Success
[Sat May 22 20:11:59 MSK 2021] Removing txt: vOY5PpdkdvOOpyix9c3QdJ07lxDkPNQOOWvoYkAA5bw for domain: _acme-challenge.le.cronfire.ru
```

**Comment by Haarolean** [comment]
Third problem:
after running `acme.sh --cron` it tries to renew the certs although they're fresh. (I remember there was a check if it's required to renew them)

**Comment by Haarolean** [comment]
Tried issuing a cert without challenge-alias:
`./acme.sh --issue -d s3.haarolean.dev --home "." --dns dns_porkbun`
The record was added for `_acme-challenge.haarolean.dev` but was checked for `s3.haarolean.dev`
`[Thu May 27 04:07:03 MSK 2021] Checking s3.haarolean.dev for _acme-challenge.s3.haarolean.dev`
I have to edit the record name manually again.

**Comment by hitchhooker** [comment]
```
satoshi@vmi565604:/etc/nginx/acme-config$ acme.sh --config-home . --issue --dns dns_porkbun -d expectchaos.com -d www.expectchaos.com
[Thu 29 Jul 2021 01:59:53 AM CEST] Using CA: https://acme.zerossl.com/v2/DV90
[Thu 29 Jul 2021 01:59:53 AM CEST] Multi domain='DNS:expectchaos.com,DNS:www.expectchaos.com'
[Thu 29 Jul 2021 01:59:53 AM CEST] Getting domain auth token for each domain
[Thu 29 Jul 2021 01:59:56 AM CEST] Getting webroot for domain='expectchaos.com'
[Thu 29 Jul 2021 01:59:56 AM CEST] Getting webroot for domain='www.expectchaos.com'
[Thu 29 Jul 2021 01:59:56 AM CEST] Adding txt value: -TimXiF76VxjJHkIZbg3LSLBQ_-v3hXjJ8YJgtDZENc for domain:  _acme-challenge.expectchaos.com
[Thu 29 Jul 2021 02:00:17 AM CEST] Error add txt for domain:_acme-challenge.expectchaos.com
[Thu 29 Jul 2021 02:00:17 AM CEST] Please add '--debug' or '--log' to check more details.
[Thu 29 Jul 2021 02:00:17 AM CEST] See: https://github.com/acmesh-official/acme.sh/wiki/How-to-debug-acme.sh
```
no idea whats wrong but cant seem to use porkbun

**Comment by Haarolean** [comment]
Run it with --debug and check the logs as it says. It will explain everything.

On Thu, Jul 29, 2021, at 03:08, Tommi Niemi wrote:
> 
> 
> ***@***.***:/etc/nginx/acme-config$ acme.sh --config-home . --issue --dns dns_porkbun -d expectchaos.com -d www.expectchaos.com
> [Thu 29 Jul 2021 01:59:53 AM CEST] Using CA: https://acme.zerossl.com/v2/DV90
> [Thu 29 Jul 2021 01:59:53 AM CEST] Multi domain='DNS:expectchaos.com,DNS:www.expectchaos.com'
> [Thu 29 Jul 2021 01:59:53 AM CEST] Getting domain auth token for each domain
> [Thu 29 Jul 2021 01:59:56 AM CEST] Getting webroot for domain='expectchaos.com'
> [Thu 29 Jul 2021 01:59:56 AM CEST] Getting webroot for domain='www.expectchaos.com'
> [Thu 29 Jul 2021 01:59:56 AM CEST] Adding txt value: -TimXiF76VxjJHkIZbg3LSLBQ_-v3hXjJ8YJgtDZENc for domain:  _acme-challenge.expectchaos.com
> [Thu 29 Jul 2021 02:00:17 AM CEST] Error add txt for domain:_acme-challenge.expectchaos.com
> [Thu 29 Jul 2021 02:00:17 AM CEST] Please add '--debug' or '--log' to check more details.
> [Thu 29 Jul 2021 02:00:17 AM CEST] See: https://github.com/acmesh-official/acme.sh/wiki/How-to-debug-acme.sh
`

**Comment by hitchhooker** [comment]
just read the topic earlier. seems like properly set config path . was the problem

**Comment by Haarolean** [comment]
Are you asking me to read my own messages again?:)) The porkbun API works fine for me with the changes from my PR #3529.

**Comment by richardharr** [comment]
i have this code:
`module.exports.getDomains = async function(apikey, config_json, domain){`
  `options = {`
    `method: 'POST',`
    `headers: {`
      `'Content-Type': 'application/json',`
      `'x-api-key': apikey`
    `},`
    `body: JSON.stringify({data: config_json})`
  `};`
  `myResponse = await fetch('https://api-ipv4.porkbun.com/api/json/v3/dns/retrieve/' + domain, options).catch(err=>{console.log(err)})`
  `domainsobject = await myResponse.json()`
  `console.log(domainsobject)`
  `return domainsobject`
`}`
this is a function that is called in a nodejs project like this:
`domainsObject = await porkbun_api.getDomains("here goes api key", config_json_object, "randomdomain")`
and it prints in the console:
{ status: 'ERROR',
  message: 'All API requests require an API key.' }
Please help!

**Comment by richardharr** [comment]
> i have this code: `module.exports.getDomains = async function(apikey, config_json, domain){` `options = {` `method: 'POST',` `headers: {` `'Content-Type': 'application/json',` `'x-api-key': apikey` `},` `body: JSON.stringify({data: config_json})` `};` `myResponse = await fetch('https://api-ipv4.porkbun.com/api/json/v3/dns/retrieve/' + domain, options).catch(err=>{console.log(err)})` `domainsobject = await myResponse.json()` `console.log(domainsobject)` `return domainsobject` `}` this is a function that is called in a nodejs project like this: `domainsObject = await porkbun_api.getDomains("here goes api key", config_json_object, "randomdomain")` and it prints in the console: { status: 'ERROR', message: 'All API requests require an API key.' } Please help!

i forgot to mention that i already tried to change inside header:
- `'x-api-key': apikey`
+ `Authorization: 'Bearer' + apikey`

**Comment by windonsky** [comment]
Used wrong api url for add txt record without "--debug". 
But with debug, it work well.

```
// without --debug, First step to remove _acme-challenge.{dom} record
_post_url='https://porkbun.com/api/json/v3/dns/retrieve/_acme-challenge.{dom}'

// The url is invalid , so get error response. Without debug, stop to add _acme-challenge.{dom} record
response='{"status":"ERROR","message":"Invalid domain."}'


// with --debug, Go to next level of {dom}
// url is valid
_post_url='https://porkbun.com/api/json/v3/dns/retrieve/{dom}'
```

I found porkbun' api about DNS Retrieve Records as:
```
// porkbun' api about DNS Retrieve Records
URI Endpoint: https://porkbun.com/api/json/v3/dns/retrieve/DOMAIN/ID

URI Endpoint: https://porkbun.com/api/json/v3/dns/retrieveByNameType/DOMAIN/TYPE/[SUBDOMAIN]

URI Endpoint: https://porkbun.com/api/json/v3/ssl/retrieve/DOMAIN
```

**Comment by allddd** [comment]
Just received the following email from Porkbun:

```
In order to ensure that any apps or tools you may have that utilize our API, we wanted to let you know about some upcoming critical updates.

We know these kinds of changes are annoying but the time has come to separate the processing of incoming API commands from our website. In previous documentation the hostname porkbun.com was used for the API but we have recently updated it to api.porkbun.com and in the near future will start enforcing its use. Please update the hostname for our API to api.porkbun.com as soon as possible.

CRITICAL UPDATE DETAILS

Type: API Hostname Change

Old Value: porkbun.com

New Value: api.porkbun.com

Deadline: 2024-12-01 00:00:00 UTC


Please note that after the deadline, API calls made to the old hostname will no longer be allowed. If you have any questions or concerns please contact support.
```

**Comment by nulluser456** [comment]
I had to add api to the front of PORKBUN_Api="https://porkbun.com/api/json/v3" in /usr/lib/acme/client/dnsapi/dns_porkbun.sh making it:

PORKBUN_Api="https://api.porkbun.com/api/json/v3"

then things started working

**Comment by neilpang** [comment]
Closing this tracking issue for now to keep the issue list clean. It remains the place to report problems with this provider -- if you hit a bug, comment here and the issue will be reopened.
